apps/src/music/views/ExemplarPlayerView.tsx

```diff
@@ -0,0 +1,152 @@
+  // Immediately load the library and source, then compile the song.
+  const onMount = useCallback(async () => {
+    workspaceRef.current.initHeadless();
+
+    workspaceRef.current.loadCode(source);
+    workspaceRef.current.compileSong(
+      {Sequencer: simple2SequencerRef.current},
+      BlockMode.SIMPLE2
+    );
+
+    // Clear any prior events and execute triggers to prepare playback.
+    simple2SequencerRef.current.clear();
+    workspaceRef.current.executeAllTriggers();
+    simple2SequencerRef.current.clear();
+    workspaceRef.current.executeCompiledSong();
+    const playbackEvents = simple2SequencerRef.current.getPlaybackEvents();
+    dispatch(setExemplarPlaybackEvents(playbackEvents));
+  }, [dispatch, source]);
+
+  useEffect(() => {
+    onMount();
+  }, [onMount]);
+
+  // Uses the already compiled song to preload sounds and play it.
+  const onPlaySong = useCallback(async () => {
+    Blockly.getMainWorkspace().hideChaff();
+    labSetPlaying(false);
+    playerRef.current?.stopSong();
+
+    const allTriggerEvents = simple2SequencerRef.current.getPlaybackEvents();
+
+    // Preload sounds and then play the song using the compiled events.
+    await playerRef.current?.preloadSounds(
+      [...allTriggerEvents, ...simple2SequencerRef.current.getPlaybackEvents()],
+      () => {}
+    );
+    playerRef.current?.playSong(
+      simple2SequencerRef.current.getPlaybackEvents()
+    );
+
+    dispatch(setPlayerContext('exemplar'));
+    setIsPlaying(true);
+  }, [dispatch, labSetPlaying]);
+
+  // Stop the exemplar song, updating Redux and local state.
+  const onStopSong = useCallback(() => {
+    if (playerContext === 'exemplar') {
+      playerRef.current?.stopSong();
+    }
+    dispatch(setPlayerContext(null));
+    setIsPlaying(false);
+  }, [playerContext, dispatch]);
+
+  useEffect(() => {
+    if (playerContext !== 'exemplar' && isPlaying) {
+      onStopSong();
+    }
+  }, [playerContext, isPlaying, onStopSong]);
+
+  const getPackDetails = (packId: string) => {
+    const packFolder = MusicLibrary.getInstance()?.getFolderForFolderId(packId);
+    if (!packFolder) {
+      return null;
+    }
+    return {
+      name: packFolder.name,
+      artist: packFolder.artist,
+      color: packFolder.color,
+      image: MusicLibrary.getInstance()?.getPackImageUrl(packId),
+    };
+  };
+
+  return (
+    <div className={moduleStyles.exemplarPlayer}>
+      <div
+        className={moduleStyles.entry}
+        key={'exemplar-player'}
+        onClick={() => {
+          isPlaying ? onStopSong() : onPlaySong();
+        }}
+      >
+        <div
+          className={classNames(
+            moduleStyles.pack,
+            isPlaying && moduleStyles.packPlaying
+          )}
+        >
+          <img
+            className={moduleStyles.packImage}
+            // The exemplar player always uses the default pack image to reduce distraction.
+            src={getPackDetails(DEFAULT_PACK)?.image}
```

Comment:
Can we just get the pack image directly instead of the full details object?

Reply:
Updated!